--- types/react/jsx-dev-runtime.d.ts
+++ types/react/ts5.0/jsx-dev-runtime.d.ts
@@ -2,7 +2,6 @@
 export { Fragment } from "./";
 
 export namespace JSX {
-    type ElementType = React.JSX.ElementType;
     interface Element extends React.JSX.Element {}
     interface ElementClass extends React.JSX.ElementClass {}
     interface ElementAttributesProperty extends React.JSX.ElementAttributesProperty {}
--- types/react/jsx-runtime.d.ts
+++ types/react/ts5.0/jsx-runtime.d.ts
@@ -2,7 +2,6 @@
 export { Fragment } from "./";
 
 export namespace JSX {
-    type ElementType = React.JSX.ElementType;
     interface Element extends React.JSX.Element {}
     interface ElementClass extends React.JSX.ElementClass {}
     interface ElementAttributesProperty extends React.JSX.ElementAttributesProperty {}
--- types/react/test/index.ts
+++ types/react/ts5.0/test/index.ts
@@ -209,6 +209,9 @@
 // allows null as props
 const FunctionComponent4: React.FunctionComponent = props => null;
 
+// undesired: Rejects `false` because of https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18051
+// leaving here to document limitation and inspect error message
+// @ts-expect-error
 const FunctionComponent5: React.FunctionComponent = () => false;
 
 // React.createElement
@@ -484,7 +487,7 @@
 // Attributes
 // --------------------------------------------------------------------------
 
-const children: any[] = ["Hello world", [null], React.createElement("span")];
+const children: any[] = ["Hello world", [null], React.createElement("span", null)];
 const divStyle: React.CSSProperties = { // CSSProperties
     flex: "1 1 main-size",
     backgroundImage: "url('hello.png')",
--- types/react/test/tsx.tsx
+++ types/react/ts5.0/test/tsx.tsx
@@ -630,15 +630,12 @@
 
 function elementTypeTests() {
     const ReturnVoid = () => {};
-    // @ts-expect-error
-    const FCVoid: React.FC = ReturnVoid;
     class RenderVoid extends React.Component {
         // @ts-expect-error
         render() {}
     }
 
     const ReturnUndefined = () => undefined;
-    const FCUndefined: React.FC = ReturnUndefined;
     class RenderUndefined extends React.Component {
         render() {
             return undefined;
@@ -646,7 +643,6 @@
     }
 
     const ReturnNull = () => null;
-    const FCNull: React.FC = ReturnNull;
     class RenderNull extends React.Component {
         render() {
             return null;
@@ -654,7 +650,6 @@
     }
 
     const ReturnNumber = () => 0xeac1;
-    const FCNumber: React.FC = ReturnNumber;
     class RenderNumber extends React.Component {
         render() {
             return 0xeac1;
@@ -662,7 +657,6 @@
     }
 
     const ReturnString = () => "Hello, Dave!";
-    const FCString: React.FC = ReturnString;
     class RenderString extends React.Component {
         render() {
             return "Hello, Dave!";
@@ -670,8 +664,6 @@
     }
 
     const ReturnSymbol = () => Symbol.for("react");
-    // @ts-expect-error
-    const FCSymbol: React.FC = ReturnSymbol;
     class RenderSymbol extends React.Component {
         // @ts-expect-error
         render() {
@@ -680,7 +672,6 @@
     }
 
     const ReturnArray = () => [<div key="one" />];
-    const FCVArray: React.FC = ReturnArray;
     class RenderArray extends React.Component {
         render() {
             return [<div key="one" />];
@@ -688,7 +679,6 @@
     }
 
     const ReturnElement = () => <div />;
-    const FCElement: React.FC = ReturnElement;
     class RenderElement extends React.Component {
         render() {
             return <div />;
@@ -696,7 +686,6 @@
     }
 
     const ReturnReactNode = ({ children }: { children?: React.ReactNode }) => children;
-    const FCReactNode: React.FC = ReturnReactNode;
     class RenderReactNode extends React.Component<{ children?: React.ReactNode }> {
         render() {
             return this.props.children;
@@ -704,6 +693,7 @@
     }
 
     const ReturnPromise = () => Promise.resolve("React");
+    // @ts-expect-error TS 5.1 only
     const FCPromise: React.FC = ReturnPromise;
     class RenderPromise extends React.Component {
         render() {
@@ -711,8 +701,8 @@
         }
     }
 
-    const ReturnPromiseReactNode = async ({ children }: { children?: React.ReactNode }): Promise<React.ReactNode> =>
-        children;
+    const ReturnPromiseReactNode = async ({ children }: { children?: React.ReactNode }) => children;
+    // @ts-expect-error TS 5.1 only
     const FCPromiseReactNode: React.FC = ReturnReactNode;
     class RenderPromiseReactNode extends React.Component<{ children?: React.ReactNode }> {
         // Undesired behavior.
@@ -745,8 +735,10 @@
     // @ts-expect-error
     React.createElement(RenderVoid);
 
-    // Desired behavior.
+    // Undesired behavior. Returning `undefined` should be accepted in all forms.
+    // @ts-expect-error
     <ReturnUndefined />;
+    // @ts-expect-error
     React.createElement(ReturnUndefined);
     <RenderUndefined />;
     React.createElement(RenderUndefined);
@@ -757,14 +749,18 @@
     <RenderNull />;
     React.createElement(RenderNull);
 
-    // Desired behavior.
+    // Undesired behavior. Returning `number` should be accepted in all forms.
+    // @ts-expect-error
     <ReturnNumber />;
+    // @ts-expect-error
     React.createElement(ReturnNumber);
     <RenderNumber />;
     React.createElement(RenderNumber);
 
-    // Desired behavior.
+    // Undesired behavior. Returning `string` should be accepted in all forms.
+    // @ts-expect-error
     <ReturnString />;
+    // @ts-expect-error
     React.createElement(ReturnString);
     <RenderString />;
     React.createElement(RenderString);
@@ -779,7 +775,10 @@
     // @ts-expect-error
     React.createElement(RenderSymbol);
 
+    // Undesired behavior. Returning `Array` should be accepted in all forms.
+    // @ts-expect-error
     <ReturnArray />;
+    // @ts-expect-error
     React.createElement(ReturnArray);
     <RenderArray />;
     React.createElement(RenderArray);
@@ -790,18 +789,23 @@
     <RenderElement />;
     React.createElement(RenderElement);
 
-    // Desired behavior.
+    // @ts-expect-error TS 5.1 only
     <ReturnReactNode />;
+    // @ts-expect-error TS 5.1 only
     React.createElement(ReturnReactNode);
     <RenderReactNode />;
     React.createElement(RenderReactNode);
 
+    // @ts-expect-error TS 5.1 only
     <ReturnPromise />;
+    // @ts-expect-error TS 5.1 only
     React.createElement(ReturnPromise);
     <RenderPromise />;
     React.createElement(RenderPromise);
 
+    // @ts-expect-error See https://github.com/microsoft/TypeScript/issues/59111
     <ReturnPromiseReactNode />;
+    // @ts-expect-error See https://github.com/microsoft/TypeScript/issues/59111
     React.createElement(ReturnPromiseReactNode);
     <FCPromiseReactNode />;
     React.createElement(FCPromiseReactNode);
--- types/react/index.d.ts
+++ types/react/ts5.0/index.d.ts
@@ -133,7 +133,7 @@
     type JSXElementConstructor<P> =
         | ((
             props: P,
-        ) => ReactNode | Promise<ReactNode>)
+        ) => ReactElement<any, any> | null)
         // constructor signature must match React.Component
         | (new(props: P, context: any) => Component<any, any>);
 
@@ -219,7 +219,7 @@
         C extends
             | ForwardRefExoticComponent<any>
             | { new(props: any, context: any): Component<any> }
-            | ((props: any) => ReactNode)
+            | ((props: any) => ReactElement | null)
             | keyof JSX.IntrinsicElements,
     > = ComponentRef<C>;
 
@@ -389,6 +389,7 @@
     }
 
     /**
+     * For internal usage only.
      * Different release channels declare additional types of ReactNode this particular release channel accepts.
      * App or library types should never augment this interface.
      */
@@ -559,7 +560,7 @@
      * @template P The props the component accepts.
      */
     interface ExoticComponent<P = {}> {
-        (props: P): ReactNode;
+        (props: P): ReactElement | null;
         readonly $$typeof: symbol;
     }
 
@@ -1049,7 +1050,7 @@
      * ```
      */
     interface FunctionComponent<P = {}> {
-        (props: P): ReactNode | Promise<ReactNode>;
+        (props: P): ReactElement<any, any> | null;
         /**
          * Ignored by React.
          * @deprecated Only kept in types for backwards compatibility. Will be removed in a future major release.
@@ -1081,9 +1082,7 @@
      *
      * @see {@link ForwardRefRenderFunction}
      */
-    // Making T nullable is assuming the refs will be managed by React or the component impl will write it somewhere else.
-    // But this isn't necessarily true. We haven't heard complains about it yet and hopefully `forwardRef` is removed from React before we do.
-    type ForwardedRef<T> = ((instance: T | null) => void) | RefObject<T | null> | null;
+    type ForwardedRef<T> = ((instance: T | null) => void) | MutableRefObject<T | null> | null;
 
     /**
      * The type of the function passed to {@link forwardRef}. This is considered different
@@ -1099,7 +1098,7 @@
      * @see {@link forwardRef}
      */
     interface ForwardRefRenderFunction<T, P = {}> {
-        (props: P, ref: ForwardedRef<T>): ReactNode;
+        (props: P, ref: ForwardedRef<T>): ReactElement | null;
         /**
          * Used in debugging messages. You might want to set it
          * explicitly if you want to display a different name for
@@ -4026,18 +4025,6 @@
 
     // Keep in sync with JSX namespace in ./jsx-runtime.d.ts and ./jsx-dev-runtime.d.ts
     namespace JSX {
-        // We don't just alias React.ElementType because React.ElementType
-        // historically does more than we need it to.
-        // E.g. it also contains .propTypes and so TS also verifies the declared
-        // props type does match the declared .propTypes.
-        // But if libraries declared their .propTypes but not props type,
-        // or they mismatch, you won't be able to use the class component
-        // as a JSX.ElementType.
-        // We could fix this everywhere but we're ultimately not interested in
-        // .propTypes assignability so we might as well drop it entirely here to
-        //  reduce the work of the type-checker.
-        // TODO: Check impact of making React.ElementType<P = any> = React.JSXElementConstructor<P>
-        type ElementType = string | React.JSXElementConstructor<any>;
         interface Element extends React.ReactElement<any, any> {}
         interface ElementClass extends React.Component<any> {
             render(): React.ReactNode;
